stonesoup/regulariser/particle.py

```diff
@@ -0,0 +1,108 @@
+            posterior particle distribution
+        detections : set of :class:`~.Detection`
+            set of detections containing clutter,
+            true detections or both
+
+        Returns
+        -------
+        particle state: :class:`~.ParticleState`
+           The particle state after regularisation
+        """
+
+        if not isinstance(posterior, ParticleState):
+            posterior = copy.copy(posterior)
+            posterior = ParticleState(None, particle_list=posterior)
+
+        if not isinstance(prior, ParticleState):
+            prior = copy.copy(prior)
+            prior = ParticleState(None, particle_list=prior)
+
+        regularised_particles = copy.copy(posterior)
+        moved_particles = copy.copy(posterior)
+
+        if detections is not None:
+            ndim = prior.state_vector.shape[0]
+            nparticles = len(posterior)
+
+            measurement_model = next(iter(detections)).measurement_model
+
+            # calculate the optimal bandwidth for the Gaussian kernel
+            hopt = (4/(ndim+2))**(1/(ndim+4))*nparticles**(-1/(ndim+4))
+            covar_est = posterior.covar
+
+            # move particles
+            moved_particles.state_vector = moved_particles.state_vector + \
+                hopt * cholesky_eps(covar_est) @ np.random.randn(ndim, nparticles)
+
+            # Evaluate likelihoods
+            part_diff = np.array(moved_particles.state_vector - prior.state_vector).T
+            part_diff_mean = np.mean(part_diff, axis=0)
+            move_likelihood = multivariate_normal.pdf(part_diff,
+                                                      mean=part_diff_mean,
+                                                      cov=np.array(covar_est))
+            post_part_diff = np.array(posterior.state_vector - prior.state_vector).T
+            post_part_diff_mean = np.mean(post_part_diff, axis=0)
+            post_likelihood = multivariate_normal.pdf(post_part_diff,
+                                                      mean=post_part_diff_mean,
+                                                      cov=covar_est)
+
+            # Evaluate measurement likelihoods
+            move_meas_likelihood = []
+            post_meas_likelihood = []
+            for detection in detections:
+                move_meas_likelihood.append(measurement_model.logpdf(detection, moved_particles))
+                post_meas_likelihood.append(measurement_model.logpdf(detection, moved_particles))
+
+            # In the case that there are multiple measurements,
+            # we select the highest overall likelihood.
+            max_likelihood_idx = np.argmax(np.sum(move_meas_likelihood, axis=1))
+
+            # Calculate acceptance probability (alpha)
+            alpha = (move_meas_likelihood[max_likelihood_idx]*move_likelihood) / \
+                    (post_meas_likelihood[max_likelihood_idx]*post_likelihood)
```

Comment:
`move_likelihood` and `post_likelihood` are from `pdf`, and `move_meas_likelihood` and `post_meas_likelihood` are using `logpdf`. Is this right?

Reply:
You are right, there is a mismatch between `logpdf` and `pdf`. I have changed the `move_likelihood` and `post_likelihood` to also be `logpdf` and changed the calculation of `alpha` accordingly.